### PR TITLE
i#6426: Add schedule_stats tool

### PR DESCRIPTION
**Pull request by derekbruening**
Adds a new analysis tool "schedule_stats" which computes characteristics of a core-sharded schedule.  This is meant to help in understanding and studying the effects of varying the scheduler parameters.

This initial version counts instructions, blocking and all syscalls, voluntary and direct switches, and direct requests, and computes derived statistics CSPKI, I/CS, %voluntary, and %direct switches. Failed direct requests can be inferred.

The tool also records a string of letters representing input threads to help visualize the thread interleaving on each core.  The number of instructions per leter is controlled by a new parameter -print_every. This string visualization has its limits as more than 26 inputs will wrap around, but this approach has been useful in our unit tests and other small runs.

Adds a regression test which runs the checked-in threadsig trace.

Issue: #6426

### CODE REVIEW

**Comment by derekbruening** [comment]
win32 and x64 failures: floatpc https://github.com/DynamoRIO/dynamorio/issues/2267, fixed by waiting-for-review https://github.com/DynamoRIO/dynamorio/pull/6441
ub22 failure: api.rseq #6185

File: clients/drcachesim/tools/schedule_stats_create.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+namespace drmemtrace {
+
+/**
+ * @file drmemtrace/schedule_stats_create.h
+ * @brief DrMemtrace schedule statistics analysis tool creation.
+ */
+
+/**
+ * Creates an analysis tool which counts the number and type of context switches
+ * in a core-sharded trace schedule.
```
Comment (abhinav92003) [inline clients/drcachesim/tools/schedule_stats_create.h:52]: Add to comment that it would fail in any other mode.

File: clients/drcachesim/tools/schedule_stats.cpp [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+            shard->cur_segment_instrs = 0;
+            shard->prev_was_wait = true;
+        } else {
+            ++shard->cur_segment_instrs;
+            if (shard->cur_segment_instrs == knob_print_every_) {
+                shard->thread_sequence += '-';
+            }
+        }
+        return true;
+    } else
```
Comment (abhinav92003) [inline clients/drcachesim/tools/schedule_stats.cpp:173]: `else` is redundant.

File: clients/drcachesim/tools/schedule_stats.cpp [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        shard->prev_was_wait = false;
```
Comment (abhinav92003) [inline clients/drcachesim/tools/schedule_stats.cpp:174]: Ideally you'd set this at the end, just in case some other logic below wants to know whether prev_was_wait.

File: clients/drcachesim/tools/schedule_stats.cpp [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    int64_t input = shard->stream->get_input_id();
+    if (input != shard->prev_input) {
+        // We convert to letters which only works well for <=26 inputs.
+        if (!shard->thread_sequence.empty()) {
+            ++shard->counters.total_switches;
+            if (shard->saw_maybe_blocking || shard->saw_exit)
+                ++shard->counters.voluntary_switches;
+            if (shard->direct_switch_target == memref.marker.tid)
+                ++shard->counters.direct_switches;
+            shard->thread_sequence += ',';
```
Comment (abhinav92003) [inline clients/drcachesim/tools/schedule_stats.cpp:184]: Maybe create separate constants up top together for all these special chars.

Comment (abhinav92003) [inline clients/drcachesim/tools/schedule_stats.cpp:184]: We add a comma at each switch? Why? isn't the character changing enough? Or do you want to differentiate between thread 0 and thread 26.

File: clients/drcachesim/tools/schedule_stats.cpp [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+                 << " == thread " << memref.instr.tid << "\n";
+            std::cerr << line.str();
+        }
+        shard->prev_input = input;
+    }
+    if (type_is_instr(memref.instr.type)) {
+        ++shard->counters.instrs;
+        // Print a single letter for any partial sequence.
+        if (shard->cur_segment_instrs == 0)
+            shard->thread_sequence += 'A' + static_cast<char>(input % 26);
```
Comment (abhinav92003) [inline clients/drcachesim/tools/schedule_stats.cpp:211]: At a thread switch L186 adds the char to the thread sequence and sets cur_segment_insrs to zero, and here we'd add the char again?

File: clients/drcachesim/common/options.h [hunk 6]
```diff
@@ -53,10 +46,20 @@
 #define VIEW "view"
 #define FUNC_VIEW "func_view"
 #define INVARIANT_CHECKER "invariant_checker"
+#define SCHEDULE_STATS "schedule_stats"
+
+#define REPLACE_POLICY_NON_SPECIFIED ""
```
Comment (abhinav92003) [inline clients/drcachesim/common/options.h:51]: Why was this moved below? If the goal was  to separate vs the tool name constants, wouldn't just an additional vertical whitespace suffice?

Also, add comment to mark the logical separation of the constants. "all tool name constants" etc.

File: clients/drcachesim/common/options.cpp [hunk 7]
```diff
@@ -854,5 +855,10 @@ droption_t<std::string>
                          "Path with stored as-traced schedule for replay.");
 #endif
 
+// Schedule_stats options.
+droption_t<uint64_t> op_print_every(DROPTION_SCOPE_ALL, "print_every", 5000,
```
Comment (abhinav92003) [inline clients/drcachesim/common/options.cpp:859]: op_schedule_stats_print_every would be better.

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/tools/schedule_stats_create.h (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/tools/schedule_stats_create.h:52]: Done.

File: clients/drcachesim/tools/schedule_stats.cpp (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/tools/schedule_stats.cpp:173]: Removed.

File: clients/drcachesim/tools/schedule_stats.cpp (hunk 3 above)
Comment (derekbruening) [reply clients/drcachesim/tools/schedule_stats.cpp:174]: Moved to the end.

File: clients/drcachesim/tools/schedule_stats.cpp (hunk 4 above)
Comment (derekbruening) [reply clients/drcachesim/tools/schedule_stats.cpp:184]: It seemed easier to read at a glance how long the sequences were.  Compare:
```
Core #0 schedule: AA,CC,HH,LL,UU,WW,YY,AA,CC,EE,FF,GG,HH,II,LL,MMM,KK,EE,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC
Core #0 schedule: AACCHHLLUUWWYYAACCEEFFGGHHIILLMMMKKEEKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCC
```
Or:
```
Core #0 schedule: FFFFFF,HHHHHHHHHHHHHHHHHHH,FFF
Core #1 schedule: DDDDDDDDDDDDDDDDDD,CCCCCCCCCCCCCCCCCCC
Core #0 schedule: FFFFFFHHHHHHHHHHHHHHHHHHHFFF
Core #1 schedule: DDDDDDDDDDDDDDDDDDCCCCCCCCCCCCCCCCCCC
```

I imagine how we visualize longer sequences will evolve over time though.  I think starting with the , is nice.  I'll add a comment about maybe removing to make it more compact, but that making it harder to distinguish a 26-rollover switch like you said.

Comment (derekbruening) [reply clients/drcachesim/tools/schedule_stats.cpp:184]: Done.

File: clients/drcachesim/tools/schedule_stats.cpp (hunk 5 above)
Comment (derekbruening) [reply clients/drcachesim/tools/schedule_stats.cpp:211]: Ah good catch.  Somehow I thought it was only adding after N so you wouldn't get a letter for a shorter sequence...not sure if the code changed in between.  But now, yes, this is redundant.  Moved this append inside the `== knob_print_every_`.

File: clients/drcachesim/common/options.h (hunk 6 above)
Comment (derekbruening) [reply clients/drcachesim/common/options.h:51]: Two reasons: A) to group it with the other per-tool defines below (CACHE_TYPE_INSTRUCTION, etc.) and B) it seemed more logical to have the tool names first, and then have per-tool specific constants.  Added comments.

File: clients/drcachesim/common/options.cpp (hunk 7 above)
Comment (derekbruening) [reply clients/drcachesim/common/options.cpp:859]: Done.

**Comment by derekbruening** [comment]
Progress: we're down to just one test failing here.  x86-32: it's the set of failures in #6417.
